### PR TITLE
Unit test coverage

### PR DESCRIPTION
**Pull request by bsweger**
@kaitlin @jworcestBAH or @nmonga91 

This request re-factors the DATA Act validator unit tests to route requests through the Flask application (instead of directly through the requests library) so that we can get an accurate baseline coverage number ([DB-196](https://federal-spending-transparency.atlassian.net/browse/DB-196).

At a high level:

* Move app creation code to a separate method in app.py so the unit tests can create a Flask instance
* Add coverage and webtest libraries
* Move setup and teardown activities to unittest setUp and tearDown methods (with a goal of laying the groundwork for a potential future use of fixtures)
* Move testUtils.py methods to a base test class
* Invoke validator functions via the Flask app
* Switch to unittest assert methods to take advantage of the built-in specific error messages
* Mark skipped tests as skipped instead of passed

A few things that were out of scope for DB-196 but should be addressed in the future:

* Tests are hitting production database names, and running them will delete existing data (see [DB-237](https://federal-spending-transparency.atlassian.net/browse/DB-237) and [DB-308](https://federal-spending-transparency.atlassian.net/browse/DB-308))
* Tests are uploading to S3. Added [DB-309](https://federal-spending-transparency.atlassian.net/browse/DB-309) as a reminder, in case this change isn't part of the impending "local" branch merge.
* There are a few hard-coded surrogate keys I'd prefer to remove when setting up test data, though given that the test data are created on empty tables, the code should work consistently in its current form.

### CODE REVIEW

**Comment by bsweger** [comment]
Resolved the two tests that were still failing when I demo'd to @kaitlin earlier today:

screen shot 2016-03-21 at 5 47 19 pm

File: requirements.txt [hunk 1]
```diff
@@ -8,5 +8,7 @@ flask-cors==2.1.2
 smart-open==1.3.1
 git+git://github.com/fedspendingtransparency/data-act-core.git@configuration
 future==0.15.2
+logstash==0.1.dev0
 python-logstash==0.4.6
-logstash==0.1.dev0
\ No newline at end of file
+coverage
+webtest
```
Comment (jworcestBAH) [inline requirements.txt]: We may want to include version numbers for coverage and webtest so we don't risk things breaking from changes in future versions of these packages.

**Comment by bsweger** [comment]
@jworcestBAH thanks for your review comments. Made some recent commits that address some of the things we spoke about:

* re-instate file size checks
* when a test fails on a status code check, add debug info to test output
* add a comment to runTests.py showing how to run a single test
* specify the webtest and coverage versions in requirements.txt

Looks like there are now some merge conflicts to resolve...will do that next.

**Comment by bsweger** [comment]
@jworcestBAH merge conflicts fixed--should be ready to go. Thanks again for the suggestions and feedback!

File: dataactvalidator/app.py [hunk 2]
```diff
@@ -118,18 +110,29 @@ def validate():
...
+    app.run(
+        debug=config["server_debug"],
+        threaded=True,
+        host=config["host"],
+        port=int(config["port"])
+    )
 
 if __name__ == "__main__" or __name__[0:5] == "uwsgi":
-    app = runApp()
+    runApp()
```
Comment (jworcestBAH) [inline dataactvalidator/app.py]: This change breaks uwsgi, I'll push a fix to make the two consistent.

File: requirements.txt [hunk 3]
```diff
@@ -7,9 +7,11 @@ Requests==2.8.1
 flask-cors==2.1.2
 smart-open==1.3.1
 future==0.15.2
-python-logstash==0.4.6
```
Comment (jworcestBAH) [inline requirements.txt]: So we actually found that order matters between the two logstash lines, this one needs to be first or you get errors when logging.  I'll push a change to restore the order.

File: tests/jobTests.py [hunk 4]
```diff
@@ -1,320 +1,264 @@
...
 
-        if not TestUtils.USE_THREADS:
-            assert(self.response.json()["message"] == "CSV file must have a header")
+        if not self.useThreads:
+            self.assertEqual(
+                response.json["message"], "CSV file must have a header")
 
     def test_missing_header(self):
-        """ Test missing header in first row """
+        """.Test missing header in first row."""
```
Comment (jworcestBAH) [inline tests/jobTests.py]: I'm not familiar with this new-fangled grammatical style of starting a sentence with a period.  Anyway, fixing this and similar in the same push as the requirements.

**Comment by jworcestBAH** [comment]
Thanks for the changes, everything looks good now and is being merged.

File: dataactvalidator/app.py (hunk 2 above)
Comment (bsweger) [inline dataactvalidator/app.py]: Thanks!

File: tests/jobTests.py (hunk 4 above)
Comment (bsweger) [inline tests/jobTests.py]: Ouch! Thanks for fixing.

**Pull request merged**